src/lib/list.c

```diff
@@ -344,6 +344,26 @@ struct list *list_deduplicate(struct list *list, comparison_fn_t comparison_fn,
 	return list;
 }
 
+struct list *list_filter_elements(struct list *list, filter_fn_t filter_fn, list_free_data_fn_t list_free_data_fn)
+{
+	struct list *iter, *next;
+
+	iter = list = list_head(list);
+	while (iter) {
+		next = iter->next;
+		if (!filter_fn(iter->data)) {
+			list_free_item(iter, list_free_data_fn);
```

Comment:
I think you are going to loose your pointer here because you are freeing the list item

Reply:
I save the pointer in next because of that.

Reply:
but in this case this condition is never going to be true `if (item == list)` because you are freeing the pointer before comparing it

Reply:
so that means that if you filter out the first element of the list list will be pointing to garbage

Reply:
When you free a pointer you don't change the value of the pointer. And you are correct that the pointer will be pointing to garbage. So accessing the value pointed by pointer will create a invalid memory access.

But in this case I'm comparing the value of the pointer that hasn't changed after a free. And this is fine. The test test_list_filter_elements() cover that in the test where I add several 'X' in the begging of the list before filtering. Try to debug that and you'll see that the value is fine and it's ok to compare them.

Reply:
ahhh I see now, thanks for the explanation